### PR TITLE
Vector aggregator api

### PR DESCRIPTION
**Pull request by aa1ex**

### CODE REVIEW

File: api/v1alpha1/vectoraggregator_types.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
```
Comment (zvlb) [inline api/v1alpha1/vectoraggregator_types.go:24]: remove

File: api/v1alpha1/vectoraggregator_types.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// VectorAggregatorSpec defines the desired state of VectorAggregator
+type VectorAggregatorSpec struct {
+	VectorCommon `json:",inline"`
+	Replicas     int32 `json:"replicas,omitempty"`
+	// Defines a filter for the Vector Pipeline and Cluster Vector Pipeline by labels.
+	// If not specified, all pipelines will be selected.
+	Selector VectorSelectorSpec `json:"selector,omitempty"`
```
Comment (zvlb) [inline api/v1alpha1/vectoraggregator_types.go:33]: pointer please

File: internal/config/configcheck/configcheck.go [hunk 3]
```diff
@@ -66,46 +67,47 @@ func New(
 	config []byte,
 	c client.Client,
 	cs *kubernetes.Clientset,
-	va *vectorv1alpha1.Vector,
+	va *vectorv1alpha1.VectorCommon,
```
Comment (zvlb) [inline internal/config/configcheck/configcheck.go:70]: vc. not va

File: internal/config/vector.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
+package config
```
Comment (zvlb) [inline internal/config/vector.go:1]: change file name to vector-source-types

File: internal/controller/pipeline_controller.go [hunk 5]
```diff
@@ -82,45 +88,116 @@ func (r *PipelineReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
...
 		}
+		for _, vector := range vectorAggregators {
+			r.VectorAggregatorsEventCh <- event.GenericEvent{Object: vector}
+		}
 		return ctrl.Result{}, nil
 	}
 
-	// Check Pipeline hash
-	checkResult, err := pipeline.CheckHash(pipelineCR)
+	pipelineChanged, err := pipeline.IsPipelineChanged(pipelineCR)
```
Comment (zvlb) [inline internal/controller/pipeline_controller.go:97]: pipelineNoChanged

File: internal/controller/pipeline_controller.go [hunk 6]
```diff
@@ -82,45 +88,116 @@ func (r *PipelineReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
...
+	if err := config.UnmarshalJson(pipelineCR.GetSpec(), p); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to unmarshal pipeline %s: %w", pipelineCR.GetName(), err)
+	}
+	pipelineVectorRole, err := p.VectorRole()
+	if err != nil {
+		if err = pipeline.SetFailedStatus(ctx, r.Client, pipelineCR, err.Error()); err != nil {
+			log.Error(err, "Failed to set pipeline status")
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
```
Comment (zvlb) [inline internal/controller/pipeline_controller.go:116]: add log for error

File: api/v1alpha1/vectoraggregator_types.go (hunk 1 above)
Comment (aa1ex) [reply api/v1alpha1/vectoraggregator_types.go:24]: done.

File: api/v1alpha1/vectoraggregator_types.go (hunk 2 above)
Comment (aa1ex) [reply api/v1alpha1/vectoraggregator_types.go]: done.

File: internal/config/configcheck/configcheck.go (hunk 3 above)
Comment (aa1ex) [reply internal/config/configcheck/configcheck.go]: done.

File: internal/controller/pipeline_controller.go (hunk 5 above)
Comment (aa1ex) [reply internal/controller/pipeline_controller.go]: fixed

File: internal/controller/pipeline_controller.go (hunk 6 above)
Comment (aa1ex) [reply internal/controller/pipeline_controller.go:117]: done.

**Review by zvlb — APPROVED** [review]
LGTM

**Pull request merged**